cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -381,12 +376,124 @@ func (m *gceManagerImpl) fetchAutoMigs() error {
 	}
 
 	if changed {
-		return m.cache.RegenerateInstancesCache()
+		return m.RegenerateInstancesCache()
 	}
 
 	return nil
 }
 
+// GetMigForInstance gets mig for given instance ref
+// Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
+// TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
+// matching prefix, but instance doesn't belong to it.
+func (m *gceManagerImpl) GetMigForInstance(instanceRef GceRef) (Mig, error) {
+	migRef, found := m.cache.GetMigRefForInstance(instanceRef)
+	if found {
+		mig, found := m.cache.getMig(migRef)
+		if !found {
+			return nil, fmt.Errorf("instance %+v belongs to unregistered mig %+v", instanceRef, migRef)
+		}
+		return mig, nil
+	}
+
+	for _, mig := range m.cache.GetMigs() {
+		migRef := mig.GceRef()
+
+		migBasename, found := m.cache.GetMigBasename(migRef)
+		var err error
+		if !found {
+			migBasename, err = m.GceService.FetchMigBasename(migRef)
+			if err != nil {
+				return nil, err
+			}
+			m.cache.SetMigBasename(migRef, migBasename)
+		}
+
+		if migRef.Project == instanceRef.Project &&
+			migRef.Zone == instanceRef.Zone &&
+			strings.HasPrefix(instanceRef.Name, migBasename) {
+
+			instanceRefs, err := m.regenerateInstanceCacheForMigReturnInstances(migRef)
+			if err != nil {
+				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", instanceRef, err)
+			}
+
+			// see if instance found among instances obatained from GCE
+			found := false
+			for _, migInstanceRef := range instanceRefs {
+				if migInstanceRef == instanceRef {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return nil, fmt.Errorf("instance %+v belongs to unknown mig", instanceRef)
+			}
+
+			mig, found := m.cache.GetMig(migRef)
+			if !found {
+				return nil, fmt.Errorf("instance %+v belongs to unregistered mig %+v", instanceRef, migRef)
+			}
+			return mig, nil
+		}
+	}
+	// Instance doesn't belong to any configured mig.
+	return nil, nil
+}
+
+func (gc *GceCache) removeInstancesForMig(migRef GceRef) {
+	for instanceRef, instanceMigRef := range gc.instanceRefToMigRef {
+		if migRef == instanceMigRef {
+			delete(gc.instanceRefToMigRef, instanceRef)
+		}
+	}
+}
+
+func (gc *GceCache) getMig(migRef GceRef) (mig Mig, found bool) {
+	mig, found = gc.migs[migRef]
+	return
+}
+
+// RegenerateInstanceCacheForMig triggers instances cache regeneration for single MIG
+func (m *gceManagerImpl) RegenerateInstanceCacheForMig(migRef GceRef) error {
```

Comment:
do we need this method? it just calls one other method

Reply:
Dropping and renaming the other one.